src/v/archival/archival_policy.cc

```diff
@@ -156,45 +186,64 @@ static upload_candidate create_upload_candidate(
+    auto ix_begin = segment->index().find_nearest(off_begin);
+    auto ix_end = segment->index().find_nearest(off_end);
+    if (
+      ix_begin.has_value() == false || ix_end.has_value() == false
+      || (segment->offsets().base_offset == off_begin && segment->offsets().dirty_offset == off_end)) {
+        // Fast path.
+        // (unlikely) We're uploading full segment because there is no offset
+        // index. Or we're uploading a full segment from the begining to the
+        // end.
         auto orig_path = std::filesystem::path(segment->reader().filename());
         vlog(
           archival_log.debug,
           "Uploading full segment {}, last_offset: {}, located segment: {}, "
           "file size: {}",
           segment->reader().filename(),
-          off,
+          off_begin,
           segment->offsets(),
           fsize);
         return {
           .source = segment,
           .exposed_name = segment_name(orig_path.filename().string()),
           .starting_offset = segment->offsets().base_offset,
+          .final_offset = segment->offsets().dirty_offset,
           .file_offset = 0,
           .content_length = fsize};
     }
-    size_t pos = ix ? ix->filepos : 0;
-    model::offset exposed_offset = ix ? ix->offset : meta->base_offset;
-    size_t clen = fsize - pos;
+    // This code path is responsible for partial upload.
+    // The partial upload is either started from arbitrary offset or ends on
+    // arbitrary offset or both.
+    // Invariant: is_end can't be std::nullopt since we checked that
+    vassert(
+      ix_begin.has_value() && ix_end.has_value(),
+      "Invariant is broken, the index can't be used.");
+    size_t pos_begin = ix_begin->filepos;
+    // The index lookup can undershoot, but if we know that the off_end is the
+    // last offset of the segment, we can compute precise file position.
+    size_t pos_end = off_end == segment->offsets().committed_offset
+                       ? segment->size_bytes()
+                       : ix_end->filepos;
+    size_t clen = pos_end - pos_begin;
     auto path = storage::segment_path::make_segment_path(
-      *ntp_conf, exposed_offset, term, version);
+      *ntp_conf, ix_begin->offset, term, version);
     vlog(
       archival_log.debug,
-      "Uploading part of the segment {}, last_offset: {}, located segment: {}, "
-      "file size: {}, starting from the offset {}",
+      "Uploading part of the segment {}, last_offset: {}, segment offsets: {}, "
+      "upload size: {}, starting from the offset {} to {}",
       segment->reader().filename(),
-      off,
+      off_begin,
       segment->offsets(),
       clen,
-      path);
+      pos_begin,
+      pos_end);
     return {
       .source = segment,
       .exposed_name = segment_name(path.filename().string()),
-      .starting_offset = exposed_offset,
-      .file_offset = pos,
+      .starting_offset = ix_begin->offset,
+      .final_offset = ix_end->offset,
```

Comment:
If the index can undershoot, can't some of the last records be left not uploaded, even if they are older than `upload_time_limit`?